### PR TITLE
krel: integrate release archiver

### PR DESCRIPTION
**Pull request by puerco**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This PR integrates the new release archiver into `krel release`. 

#### Which issue(s) this PR fixes:

Part of #1673

Depends on ~#1768~

#### Special notes for your reviewer:

Tested and ready :rocket: 

#### Does this PR introduce a user-facing change?
```release-note
- Integrate the release archiver into `krel release`
```

### CODE REVIEW

File: pkg/anago/release.go [hunk 1]
```diff
@@ -419,4 +426,20 @@ func (d *DefaultRelease) CreateAnnouncement() error {
...
+// its logs for long term conservation
+func (d *DefaultRelease) Archive() error {
+	// Create a new options set for the release archiver
+	archiverOptions := &release.ArchiverOptions{
+		ReleaseBuildDir: filepath.Join(workspaceDir, d.options.BuildVersion),
+		LogsDirectory:   filepath.Join(workspaceDir, "tmp"),
+		StageGCSPath:    "gs://" + filepath.Join(d.options.Bucket(), "stage"),
+		ArchiveGCSPath:  "gs://" + filepath.Join(d.options.Bucket(), "archive"),
+		BuildVersion:    d.options.BuildVersion,
+	}
```
Comment (puerco) [inline pkg/anago/release.go:439]: :warning:  I want to double check these values.

From the logs and the code, I think the paths inside the GCB workspace directory are different in `krel` than in `anago`. The values here are from the anago environment, let's talk about this.

Comment (saschagrunert) [reply pkg/anago/release.go:439]: Do we even write the logs right now? We probably have to tell logrus to write to a file (beside to stderr). 

The workspace is hard-coded to `/workspace`, where everything else derives from it (like `gitRoot`).

File: pkg/anago/release.go [hunk 2]
```diff
@@ -419,4 +426,20 @@ func (d *DefaultRelease) CreateAnnouncement() error {
...
 
-func (d *DefaultRelease) Archive() error { return nil }
+// Archive stores the release artifact in a bucket along with
+// its logs for long term conservation
+func (d *DefaultRelease) Archive() error {
+	// Create a new options set for the release archiver
+	archiverOptions := &release.ArchiverOptions{
+		ReleaseBuildDir: filepath.Join(workspaceDir, d.options.BuildVersion),
+		LogsDirectory:   filepath.Join(workspaceDir, "tmp"),
+		StageGCSPath:    "gs://" + filepath.Join(d.options.Bucket(), "stage"),
```
Comment (saschagrunert) [inline pkg/anago/release.go:436]: let's use `object.GcsPrefix` for this.

File: pkg/anago/release.go (hunk 1 above)
Comment (puerco) [reply pkg/anago/release.go:439]: > Do we even write the logs right now? 

We do not write them, I thought about that and assumed we would write them at at some point to emulate what anago did. 

> The workspace is hard-coded to /workspace, where everything else derives from it (like gitRoot).

This is precisely what I wanted to confirm. From what I can understand, in anago, the build happened under a directory named with the version, for example gitRoot then was:

`/workspace/anago-v1.20.0-beta.2/src/`

Now, the version subdirectory is gone and the build happens one level up, directly in `/workspace`. So the git root is now just:

`/workspace/src/`

Am I right?

Comment (saschagrunert) [reply pkg/anago/release.go:439]: Yep, that's right.

Created a PR for the logs https://github.com/kubernetes/release/pull/1774 (they will be put into $TMPDIR/{stage,release}.log and can be retrieved from the `state`)

Comment (puerco) [reply pkg/anago/release.go:439]: Awesome, I'll patch the archive step to collect them from the new names/locations.

File: pkg/anago/release.go (hunk 2 above)
Comment (puerco) [reply pkg/anago/release.go]: These options are gone, but internally i'm using the constant

File: pkg/anago/release.go [hunk 3]
```diff
@@ -444,4 +451,20 @@ func (d *DefaultRelease) CreateAnnouncement() error {
...
+// its logs for long term conservation
+func (d *DefaultRelease) Archive() error {
+	// Create a new options set for the release archiver
+	archiverOptions := &release.ArchiverOptions{
+		ReleaseBuildDir: filepath.Join(workspaceDir, "src"),
+		LogFile:         d.state.logFile,
+		BuildVersion:    d.options.BuildVersion,
+		PrimeVersion:    d.state.versions.Prime(),
+		Bucket:          d.options.Bucket(),
+	}
```
Comment (puerco) [inline pkg/anago/release.go:464]: I've simplified the archiver interface.

Comment (saschagrunert) [reply pkg/anago/release.go:464]: Looks good!

**Review by saschagrunert — APPROVED** [review]
/lgtm

**Comment by saschagrunert** [comment]
@puerco feel free to change the PR title to remove the `WIP` to let this change get merged.

**Comment by puerco** [comment]
#### Test Runs:
:hourglass_flowing_sand: **Stage:** https://console.cloud.google.com/cloud-build/builds/4f7bf2d1-57a5-4c6d-b65b-57fb4c4d8e0c?authuser=1&project=kubernetes-release-test
:hourglass: Release:

**Comment by puerco** [comment]
Rebased to let #1782 into my branch.

File: pkg/release/archive_unit_test.go [hunk 4]
```diff
@@ -23,6 +23,7 @@ import (
 	"testing"
 
 	"github.com/stretchr/testify/require"
+	"k8s.io/release/pkg/object"
```
Comment (saschagrunert) [inline pkg/release/archive_unit_test.go:26]: nit:
```suggestion
	"github.com/stretchr/testify/require"

	"k8s.io/release/pkg/object"
```

File: pkg/anago/release.go [hunk 5]
```diff
@@ -244,6 +245,12 @@ func (d *defaultReleaseImpl) CreateAnnouncement(options *announce.Options) error
 	return announce.CreateForRelease(options)
 }
 
+func (d *defaultReleaseImpl) ArchiveRelease(options *release.ArchiverOptions) error {
+	// Create a new release archiver
+	archiver := release.NewArchiver(options)
+	return archiver.ArchiveRelease()
```
Comment (saschagrunert) [inline pkg/anago/release.go:251]: nit:
```suggestion
	return release.NewArchiver(options).ArchiveRelease()
```

File: pkg/release/archive.go [hunk 6]
```diff
@@ -84,29 +85,35 @@ func (o *ArchiverOptions) ArchiveBucketPath() string {
...
 	// Check if the tag is well formed
 	_, err := util.TagStringToSemver(o.BuildVersion)
 	if err != nil {
-		return errors.Wrap(err, "verifying release tag")
+		return errors.Wrap(err, "verifying build version tag")
+	}
+
+	// Check if the tag is well formed
+	_, err = util.TagStringToSemver(o.PrimeVersion)
+	if err != nil {
```
Comment (saschagrunert) [inline pkg/release/archive.go:115]: nit:
```suggestion
	if_, err = util.TagStringToSemver(o.PrimeVersion); err != nil {
```

**Comment by puerco** [comment]
#### Test runs round 2
:heavy_check_mark:  (Clean) Release test run with nits, log transfer at the end and rsync instead of copy: https://console.cloud.google.com/cloud-build/builds/037eb543-55ad-4070-8b33-6eee998c13e4?project=kubernetes-release-test
:hourglass_flowing_sand:  Second release test run with rsync to bucket: https://console.cloud.google.com/cloud-build/builds/6345aeb0-c9c6-4367-9af3-b93a95a96f3d?project=kubernetes-release-test

**Review by saschagrunert — APPROVED** [review]
/lgtm

**Pull request merged**